renderer/brave_wallet/brave_wallet_render_frame_observer.cc

```diff
@@ -45,9 +45,16 @@ void BraveWalletRenderFrameObserver::DidCreateScriptContext(
     js_ethereum_provider_.reset();
     return;
   }
+
   // Wallet provider objects won't be generated for third party iframe
-  if (!render_frame()->IsMainFrame() &&
-      render_frame()->GetWebFrame()->IsCrossOriginToOutermostMainFrame()) {
+  if ((!render_frame()->IsMainFrame() &&
+       render_frame()->GetWebFrame()->IsCrossOriginToOutermostMainFrame()) ||
+      (!render_frame()->IsMainFrame() && !render_frame()
+                                              ->GetMainRenderFrame()
+                                              ->GetWebFrame()
+                                              ->GetDocument()
+                                              .IsSecureContext()) ||
+      !render_frame()->GetWebFrame()->GetDocument().IsSecureContext()) {
```

Comment:
IIUC this is checking that the iframe is secure and the top level frame is secure; however it misses the case where the frame chain is `secure context -> insecure context -> secure context`. so the provider will be injected in a nested secure iframe embedded in an insecure iframe embedded in a secure top-level frame.

Chromium must have logic for checking this case somewhere since it has to do so for geolocation - can we simply reuse that logic?

Reply:
probably something like this: https://source.chromium.org/chromium/chromium/src/+/main:content/browser/renderer_host/navigation_request.cc;l=283?q=issecureframe&sq=&ss=chromium

Reply:
actually maybe frame ancestors are already taken into account in `IsSecureContext()` https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/execution_context/security_context.h;l=65;drc=18028015e9e3f7085d20882a71148a3b81e2034e;bpv=1;bpt=1

Reply:
well the origin would be different for that because you can't have both HTTP and HTTPS serving from the same port. Currently we check strictly for the same origin.

Reply:
That cross origin check is only to the outer frame though.

Reply:
@diracdeltas OK I did a test and I think it's right as is. Your comment here is correct
https://github.com/brave/brave-core/pull/13739/files#r894896883



Checked by setting:
brave://flags/#unsafely-treat-insecure-origin-as-secure

To http://z.com:8080

Then I started a local server.
Both z.com and y.com mapped to 127.0.01 in the hosts file.

z.com main frame -> z.com iframe -> z.com iframe
They all have window.isSecureContext and window.ethereum is defined.

z.com main frame -> y.com iframe -> z.com iframe
Only on the top level is it a secure context and window.ethereum is defined.  Both iframes have insecure context and no window.ethereum.

Reply:
I removed the redundant check for `(!render_frame()->IsMainFrame() && !render_frame()
                                              ->GetMainRenderFrame()
                                              ->GetWebFrame()
                                              ->GetDocument()
                                              .IsSecureContext())` since based on those findings I just posted it is not needed.

Reply:
sgtm, thanks